### PR TITLE
Adding Openshift stack

### PR DESCRIPTION
**Pull request by nakfour**
**Which issue is resolved by this Pull Request:**
This is to upgrade all changes needed to run Kubeflow on OCP 4.x to kustomize v3.
Resolves #

**Description of your changes:**

**Checklist:**
- [ ] Unit tests have been rebuilt: 
    1. `cd manifests/tests`
    2. `make generate-changed-only`
    3. `make test`

### CODE REVIEW

**Review by crobby — APPROVED** [review]

**Review by crobby — COMMENTED** [review]
Looks like its on the right path.  Great work!

**Comment by vpavlin** [comment]
Metadata fail to start because metadata-db livenessprobe is failing. The liveness probe needs to use `MYSQL_USER` to check mysql availability. You can use my commit here:

https://github.com/vpavlin/manifests/commit/d9bce160281988f8e498a96a8f9633e479670b6a

**Comment by nakfour** [comment]
@vpavlin now I get this error below in on of the metadata-db pods that is showing up "not ready" then keeps crashing
```
[ERROR] InnoDB: Unable to lock ./ibdata1 error: 11
```

**Comment by nakfour** [comment]
@jlewi @animeshsingh is there a way to get detailed logs from the kubeflow-manifests-presubmit job? Its failing and we have no indication why?

**Comment by Jeffwan** [comment]
@nakfour 

Did you run test locally? 

seems Argo component doesn't load logs correctly. 

http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-manifests-presubmit-unit-1567-02bd38e-1376-0d7e?tab=workflow&nodeId=kubeflow-manifests-presubmit-unit-1567-02bd38e-1376-0d7e-202129096&sidePanel=logs%3Akubeflow-manifests-presubmit-unit-1567-02bd38e-1376-0d7e-202129096%3Amain

**Comment by Jeffwan** [comment]
If you plan to have this in v1.2 release, please let us know https://github.com/kubeflow/kubeflow/issues/5224

**Comment by nakfour** [comment]
@Jeffwan yes ran test manually and ran into issues documented here : https://github.com/kubeflow/manifests/issues/1596
I do not see the argo issue you are referring to, please point it out?

File: kfdef/kfctl_openshift.v1.1.0.yaml
Comment (Tomcli) [inline kfdef/kfctl_openshift.v1.1.0.yaml:12]: should we change the `application` directory name to plural similar to `stacks`?

File: kfdef/kfctl_openshift.v1.1.0.yaml
Comment (Tomcli) [inline kfdef/kfctl_openshift.v1.1.0.yaml:89]: Is it possible to install `pipeline` and the Kubeflow control plane separately? Right now in this kfdef, I don't see a way to opt-out pipeline like other `application`.

File: kfdef/kfctl_openshift.v1.1.0.yaml
Comment (Tomcli) [reply kfdef/kfctl_openshift.v1.1.0.yaml:12]: Discard this comment, somehow the singular application is being used on other stacks as well.

File: kfdef/kfctl_openshift.v1.1.0.yaml
Comment (nakfour) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: @Tomcli we would like to do that, however we tried and it did not work. We documented the issue here in a comment: https://github.com/nakfour/manifests/blob/openshift_stack/stacks/openshift/kustomization.yaml
and here https://github.com/opendatahub-io/manifests/issues/54
In summary, pipeline needs profiles setup first, if we take profiles as a single application and run it, $namespace for the virtualservice is not being replaced, so it fails. The plan is to keep it in the main app and re-visit this later.

Comment (Tomcli) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: Technically you can name two applications with the same name. This way kfctl will build those two applications into a single kustomization and shares the same variables. However, you will need to use the kfctl 1.2 (master branch) because kfctl 1.1 has few bugs with the same application name.

Comment (nakfour) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: @Tomcli not sure I understand your comment. Which two applications have the same name?

Comment (moficodes) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: @nakfour https://github.com/kubeflow/manifests/blob/master/kfdef/kfctl_ibm_customstack.yaml#L62-L112 
all these are part of the same application

Comment (Tomcli) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: To clarify, when you put the same name on two applications, kfctl build will group those two applications into a single kustomize. Here you can put `pipelines` and `base` to have the same applications.name `kubeflow-apps`, so they can share the `$namespace` variable.

Comment (animeshsingh) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: @nakfour have requested @Tomcli and @moficodes to review your PR so that we can adopt a consistent style across RedHat and IBM minimum, and it will be easier for us to get our customized changes in

**Comment by animeshsingh** [comment]
> @jlewi @animeshsingh is there a way to get detailed logs from the kubeflow-manifests-presubmit job? Its failing and we have no indication why?

@nakfour can you rebase your PR on the latest and rerun. Hopefully that clears you up

**Comment by Tomcli** [comment]
For the tests I opened a quick PR #1604 to fix it, it looks like we are using a very old kyaml library in our tests and it cannot load non SequenceNode yaml.

**Comment by animeshsingh** [comment]
thanks @Tomcli !

**Comment by animeshsingh** [comment]
@nakfour please rebase once, and hopefully this shall be good.

**Comment by nakfour** [comment]
Thanks @Tomcli and @animeshsingh, rebased, the tests skip the non SequenceNode, but still fail on code outside our stack
```
=== RUN   TestValidK8sResources
--- FAIL: TestValidK8sResources (0.98s)
    validate_resources_test.go:249: Path ../istio/istio-install/base/istio-install-test.yaml; resource jaeger-query; has empty annotations; if no annotations are present the field shouldn't be present
    validate_resources_test.go:249: Path ../istio/istio-install/base/istio-install-test.yaml; resource tracing; has empty annotations; if no annotations are present the field shouldn't be present
    validate_resources_test.go:236: Path ../jupyter/notebook-controller/base/notebook-controller-test.yaml; resource notebooks.kubeflow.org; has status field
```

**Comment by Tomcli** [comment]
> Thanks @Tomcli and @animeshsingh, rebased, the tests skip the non SequenceNode, but still fail on code outside our stack
> 
> ```
> === RUN   TestValidK8sResources
> --- FAIL: TestValidK8sResources (0.98s)
>     validate_resources_test.go:249: Path ../istio/istio-install/base/istio-install-test.yaml; resource jaeger-query; has empty annotations; if no annotations are present the field shouldn't be present
>     validate_resources_test.go:249: Path ../istio/istio-install/base/istio-install-test.yaml; resource tracing; has empty annotations; if no annotations are present the field shouldn't be present
>     validate_resources_test.go:236: Path ../jupyter/notebook-controller/base/notebook-controller-test.yaml; resource notebooks.kubeflow.org; has status field
> ```

You need to use kustomize v3.2.1 since that's the kustomize version for generating all these tests. You can see that the CI passes without these issues.

**Comment by nakfour** [comment]
@Tomcli I am using Kustomize v3.4.0

**Comment by Tomcli** [comment]
> @Tomcli I am using Kustomize v3.4.0

From https://github.com/kubeflow/manifests#manifests
Anything above Kustomize v3.2.1 may have problems with the tests.

File: kfdef/kfctl_openshift.v1.1.0.yaml
Comment (nakfour) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: @animeshsingh @Tomcli looking at kfctl_ibm.v1.1.0.yaml each app has a different name, same for kfctl_k8s_istio.v1.1.0.yaml. I am suggesting we keep kfctl_openshift.v1.1.0.yaml as it is now. 
/cc @vpavlin @crobby

Comment (vpavlin) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: @Tomcli having the same name for multiple apps sounds like a dirty workaround to an actual problem - like sharing a set of common/global variables/parameters - are there any plans to resolve this more systemically?

One feature I have in mind for the operator is to set state (success/failure) for each application name in `status` of the CR - this would allow users to see how the deployment went - obviously how do we figure out if the deployment was successful is a separate question..but if all the apps have the same name we lose this option, so I don't think this is a good idea in general

File: stacks/openshift/application/katib/katib-mysql-deployment.yaml
Comment (vpavlin) [inline stacks/openshift/application/katib/katib-mysql-deployment.yaml:11]: missing new line

File: kfdef/kfctl_openshift.v1.1.0.yaml
Comment (Tomcli) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: As of now, kfctl doesn't have a way to define global variables. I opened an issue at https://github.com/kubeflow/kfctl/issues/429 to address this issue. It's fine to keep the `kfctl_openshift.v1.1.0.yaml` as it for now, but it would be nice to have some comments in the kfdef to clarify what's in the base application.

**Comment by PatrickXYS** [comment]
Can you try re-run new test?

/test ?

File: kfdef/kfctl_openshift.v1.1.0.yaml
Comment (animeshsingh) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: @vpavlin @nakfour @Tomcli @moficodes is there an other alternative way in lieu of this to be consistent? What we don't like is bundling together something like pipelines with base stack in core openshift stacj, which means for someone just looking at pipelines they will be sort of forced to get the core components as well..... We ideally want to figure out an easier way to swap-in/swap out pipelines from OpenShift manifests, and hence aligning on a consistent style across will make more sense, else its a combination of multiple kfdef files floating for OpenShift

Comment (animeshsingh) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: > One feature I have in mind for the operator is to set state (success/failure) for each application name in `status` of the CR - this would allow users to see how the deployment went - obviously how do we figure out if the deployment was successful is a separate question..

@vpavlin @nakfour if this is something you can drive at your end, believe would be helpful for the overall odh umbrella as well. Not sure of the effort required in terms of the timeline et all.

**Comment by PatrickXYS** [comment]
/hold

Hold until rebase and re-run test

File: kfdef/kfctl_openshift.v1.1.0.yaml
Comment (nakfour) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: @animeshsingh I agree we need a way to allow users to add/remove pipelines, we need to table this discussion for next release. Please open an issue for it and we can track.

**Comment by nakfour** [comment]
@PatrickXYS rebase done
/test all

File: kfdef/kfctl_openshift.v1.1.0.yaml
Comment (animeshsingh) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: Thanks @nakfour 
Next release of what?

**Comment by crobby** [comment]
I'm in the process of trying this out on OpenShift 4.6 and it looks like Seldon will not work due to a known issue:  https://github.com/kubeflow/manifests/issues/1543

**Comment by nakfour** [comment]
@crobby the seldon issue should be fixed in the upgrade: https://github.com/kubeflow/manifests/pull/1600

File: kfdef/kfctl_openshift.v1.1.0.yaml
Comment (Tomcli) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: @nakfour I think it's possible to add/remove pipelines as an independent application. I see there's some conflicts on putting multi-user kfp on the single user openshift settings. Also there's a bug on the scc settings that's probably failing the deployment.

After apply the below fixes, we are able to deploy your openshift kfdef on OCP 4.3 with the pipeline application. On OCP 4.5 only seldon is failing but it should be resolved in #1600 after the merge.
https://github.com/nakfour/manifests/compare/openshift_stack...Tomcli:openshift-refactor

Comment (Tomcli) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: I do think able to add/remove pipelines is important because users can have an easier time to switch between single and multi-user kfp.

**Comment by crobby** [comment]
I think notebooks spin-up is also failing due to the fact it's trying to use user/group ids that are not permitted in OpenShift.

**Comment by nakfour** [comment]
@crobby istio side car injection requires  NET_ADMIN capability as described in this blog: https://www.openshift.com/blog/increasing-security-of-istio-deployments-by-removing-the-need-for-privileged-containers
in order to change the ip tables. So in OCP we have to give each pod privileged access in that namespace which is not recommended. I will try and see if I can disable side car injection.

File: kfdef/kfctl_openshift.v1.1.0.yaml
Comment (animeshsingh) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: @vpavlin @nakfour would be good to resolve this before we merge this PR. I have mentioned many times we are currently the biggest user for KF on OpenShift, and if we want consistency and not wanting multiple KF on OpenShift instructions mechanisms and kfdefs to proliferate, its best we align on one way. Additionally, I am not sure if you are planning to launch KF 1.2 on OpenShift, or KF 1.1.  If it's KF 1.1 we can wait a while, resolve and align before we get this in.

Comment (vpavlin) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: > We ideally want to figure out an easier way to swap-in/swap out pipelines from OpenShift manifests,

Can we first try to get this merged and then revisit this, as @nakfour suggested? I understand your priorities, but I think it is worth having this in ASAP and then refactor in smaller chunks

> if this is something you can drive at your end,

Yes, that is the plan, we need this for our deployments internally and on MOC

> I am not sure if you are planning to launch KF 1.2 on OpenShift

The last we talked about this, I believe the agreement was we'd skip 1.1 and target 1.2 instead to catch up with KF....Do I remember correctly, @nakfour?

Comment (Tomcli) [reply kfdef/kfctl_openshift.v1.1.0.yaml:89]: > Can we first try to get this merged and then revisit this, as @nakfour suggested? I understand your priorities, but I think it is worth having this in ASAP and then refactor in smaller chunks

I found a few bugs in the openshift/application. https://github.com/nakfour/manifests/compare/openshift_stack...Tomcli:openshift-refactor can provide an easier way to swap-in/swap out single user pipelines.

File: stacks/openshift/application/openshift/openshift-scc/kubeflow-anyuid-scc-istio.yaml
Comment (Tomcli) [inline stacks/openshift/application/openshift/openshift-scc/kubeflow-anyuid-scc-istio.yaml:45]: missing istio-system namespace, without it Kubeflow cannot communicate with the cluster-local-gateway
```suggestion
- system:serviceaccount:istio-system:cluster-local-gateway-service-account
```

File: stacks/openshift/application/pipeline/kustomization.yaml
Comment (Tomcli) [inline stacks/openshift/application/pipeline/kustomization.yaml:7]: Why are we using multi-user kfp over here? In this PR's kfdef, there's no authentication services and istio sidecar injection to provide istio auth. Thus, users cannot run pipelines on the multi-user kfp because they cannot authenticate themselves.

Comment (Tomcli) [reply stacks/openshift/application/pipeline/kustomization.yaml:7]: Here is how to make it run with single user, https://github.com/nakfour/manifests/compare/openshift_stack...Tomcli:openshift-refactor

File: stacks/openshift/application/openshift/openshift-scc/kubeflow-anyuid-scc-istio.yaml
Comment (nakfour) [reply stacks/openshift/application/openshift/openshift-scc/kubeflow-anyuid-scc-istio.yaml]: @Tomcli good catch, thanks.

File: stacks/openshift/application/pipeline/kustomization.yaml
Comment (nakfour) [reply stacks/openshift/application/pipeline/kustomization.yaml:7]: We disabled istio-injection, as posted in a comment before. We at the moment cannot pull pipeline into applications because of its dependency on profiles. We discussed this before

Comment (Tomcli) [reply stacks/openshift/application/pipeline/kustomization.yaml:7]: Yes, I understand you want to disable istio-injection, so we should using the single user kfp isn't it? With the single user kfp, there shouldn't be a dependency on profiles.

Comment (Tomcli) [reply stacks/openshift/application/pipeline/kustomization.yaml:7]: `../../../../pipeline/installs/generic` is the single user mode kfp

Comment (nakfour) [reply stacks/openshift/application/pipeline/kustomization.yaml:7]: Ok but. we would like users to run pipelines in their own namespace generated by the profile controller. Isn't that the whole point of multi-user pipelines

**Comment by nakfour** [comment]
@PatrickXYS @Tomcli  how can we get logs on what is failing in the e2e test? Running tests locally pass, and this stack is specific to Openshift cluster.

File: stacks/openshift/application/pipeline/kustomization.yaml
Comment (Tomcli) [reply stacks/openshift/application/pipeline/kustomization.yaml:7]: Sure, but in that case you will need to provide authentication services and istio sidecar injection to authenticate incoming user traffics. And the current istio sidecar injection is not ideal for openshift.

Comment (Tomcli) [reply stacks/openshift/application/pipeline/kustomization.yaml:7]: right now with this deployment, you cannot submit pipelines as a user.

Comment (nakfour) [reply stacks/openshift/application/pipeline/kustomization.yaml]: Yes switched pipeline to generic and pulled out into kfdef

**Comment by Tomcli** [comment]
thanks @nakfour 
/lgtm

**Comment by animeshsingh** [comment]
Thanks @nakfour 
/lgtm
/approve

**Pull request merged**
